packages/fuels-core/src/code_gen/custom_types_gen.rs

```diff
@@ -258,22 +259,41 @@ pub fn expand_custom_enum(enum_name: &str, prop: &Property) -> Result<TokenStrea
                 param_types.push(quote! { types.push(ParamType::Unit) });
                 args.push(quote! {(#dis, token, _) => Ok(#enum_ident::#variant_name()),});
             }
-            // Elementary type
+            // Elementary types, String, Array
             _ => {
                 let ty = expand_type(&param_type)?;
```

Comment:
Tried to make this a bit cleaner:
```rust
let param_type_construction = gen_param_type_constructing_code(&param_type)?;
param_types.push(quote! { types.push(#param_type_construction) });

let ty = expand_type(&param_type)?;
enum_variants.push(quote! { #variant_name(#ty)});

args.push(quote! {(#dis, token, _) => Ok(#enum_ident::#variant_name(<#ty>::from_token(token)?)),},);

let token = gen_token_constructing_code(&param_type);
enum_selector_builder.push(quote! {
  #enum_ident::#variant_name(value) => (#dis,  #token)
 });
```
And the helpers:
```rust
fn gen_param_type_constructing_code(param_type: &ParamType) -> Result<TokenStream, LexError> {
    let variant = TokenStream::from_str(&param_type.to_string())?;
    Ok(quote! {ParamType::#variant})
}
fn gen_token_constructing_code(param_type: &ParamType) -> TokenStream {
    let init_value = if let ParamType::Array(..) = param_type {
        quote! { vec![value.into_token()] }
    } else {
        // Primitive type
        quote! { value }
    };

    let token_type = determine_token_variant(&param_type);
    quote! {Token::#token_type(#init_value)}
}
fn determine_token_variant(param_type: &ParamType) -> Ident {
    let token_variant = match param_type {
        ParamType::Array(..) => "Array".to_string(),
        ParamType::String(..) => "String".to_string(),
        _ => param_type.to_string(),
    };
    ident(&token_variant)
}
```

Even more could be achieved in terms of readability but it would require more refactoring outside the scope of your changes.

Reply:
Now that you mention it, it does seem like this could be generalized. The ones above 100% work only for enums, didn't check if the other cases have anything different with regards to the parameters and tokens construction.

Reply:
Yes that's what I was hinting at in my previous comment. It's definitely a good tech debt to fix, but it should be done in a separate PR.